tutorial/tests/testsuite.py

```diff
@@ -49,9 +59,139 @@ def pytest_generate_tests(self, metafunc: pytest.Metafunc) -> None:
             metafunc.parametrize("function_to_test", [self.function_to_test])
 
 
+@dataclasses.dataclass
+class TestResult:
+    """Container class to store the test results when we collect them"""
+    stdout: str
+    stderr: str
+    test_name: str
+    success: bool
+
+
+@dataclasses.dataclass
+class OutputConfig:
+    """Container class to store the information to display in the test output"""
+    style: str
+    name: str
+    result: str
+
+
+def format_success_failure(syntax_error: bool, success: bool, name: str) -> OutputConfig:
+    """
+    Depending on the test results, returns a fragment that represents
+    either an error message, a success message, or a syntax error warning
+    """
+    if syntax_error:
+        return OutputConfig(
+            "alert-warning",
+            f"Tests <strong>CANNOT RUN</strong> for this cell.",
+            "&#129300 Careful, looks like you have a syntax error.",
+        )
+
+    if success:
+        return OutputConfig(
+            "alert-success",
+            f"Tests <strong>PASSED</strong> for the function <code>{name}</code>",
+            "&#x1F64C Congratulations, your solution was correct!",
+        )
+    else:
+        return OutputConfig(
+            "alert-danger",
+            f"Tests <strong>FAILED</strong> for the function <code>{name}</code>",
+            "&#x1F631 Your solution was not correct!",
+        )
+
+
+def format_long_stdout(text: str) -> str:
+    """
+    Format a long test stdout as a HTML by using the <details> element
+    """
+
+    def join_lines(text: str) -> str:
+        text = text.splitlines()
+        return "".join(
+            [f"<p>{line}</p>" for line in text[-6:-1]]
+        )
+
+    separator = 59 * ' _'
+    text = text.split(separator)[-1]
+    text_lines = "".join(join_lines(text))
+    test_runs = "".join(
+        [
+            f"""<details style="overflow-y: auto; max-height: 200px;"><summary><u>Click here to expand</u></summary><div>{text_lines}</div></details></li>"""
+        ]
+    )
```

Comment:
I find these lines a bit confusing: `join_lines` returns a string, so why on line 118 we are joining again? It doesn't do anything, or am I missing something?

I think all this can be rewritten in a clearer/more compact way. **To be tested**!

```suggestion
    def join_lines(text: str) -> str:
        text = text.splitlines()
        return "".join(
            f"<p>{line}</p>" for line in text[-6:-1]
        )

    separator = 59 * ' _'
    text = text.split(separator)[-1]
    
    test_runs =  f"""<details style="overflow-y: auto; max-height: 200px;"><summary><u>Click here to expand</u></summary><div>{join_lines(text)}</div></details></li>"""
    )
```